test/core/ExamplesUnitTests/CMakeLists.txt

```diff
@@ -841,6 +841,228 @@ TRIBITS_ADD_ADVANCED_TEST( TribitsHelloWorld_install_config_dummy_proj
+  # nvcc_wrapper.)
+
+
+TRIBITS_ADD_ADVANCED_TEST( TribitsHelloWorld_no_ctest_resources_file
+  OVERALL_WORKING_DIRECTORY TEST_NAME
+  OVERALL_NUM_MPI_PROCS 1
+  EXCLUDE_IF_NOT_TRUE HAS_CTEST_RESOURCE_FILE_SUPPORT
+
+  TEST_0
+    MESSAGE "Configure without any ctest resources file, just enable CUDA"
+    CMND ${CMAKE_COMMAND}
+    ARGS
+      ${TribitsHelloWorld_COMMON_CONFIG_ARGS}
+      -DTribitsHelloWorld_ENABLE_TESTS=ON
+      -DTPL_ENABLE_CUDA=ON
+      ${${PROJECT_NAME}_TRIBITS_DIR}/examples/TribitsHelloWorld
+    PASS_REGULAR_EXPRESSION_ALL
+      "Configuring done"
+    ALWAYS_FAIL_ON_NONZERO_RETURN
+
+  TEST_1
+    MESSAGE "Check that the CTestTestfile.cmake file has the right entries"
+    CMND cat
+    ARGS hello_world/CTestTestfile.cmake
+    PASS_REGULAR_EXPRESSION_ALL
+      "ENVIRONMENT .CTEST_KOKKOS_DEVICE_TYPE=gpus."
+      "PROCESSORS .1."
+      "RESOURCE_GROUPS .1,gpus:1."
+    ALWAYS_FAIL_ON_NONZERO_RETURN
+
+  )
+  # NOTE: The above test ensures that the RESOURCE_GROUPS test property gets
+  # set correctly in a CUDA build of a project, regardless if a ctest
+  # resources file gets set or not.  (The user can always explicitly pass in a
+  # ctest resources file later directly to ctest.)
+
+
+TRIBITS_ADD_ADVANCED_TEST( TribitsHelloWorld_set_other_ctest_resources_file_autogen_on
+  OVERALL_WORKING_DIRECTORY TEST_NAME
+  OVERALL_NUM_MPI_PROCS 1
+  EXCLUDE_IF_NOT_TRUE HAS_CTEST_RESOURCE_FILE_SUPPORT
+
+  TEST_0
+    MESSAGE "Copy in pre-created ctest resource file to make sure it is not overwritten"
+    CMND cp
+    ARGS "${CMAKE_CURRENT_SOURCE_DIR}/ctest_resources/other_ctest_resources.json" .
+
+  TEST_1
+    MESSAGE "Configure pointing to non-default ctest resource file with autogenerate turned on "
+    CMND ${CMAKE_COMMAND}
+    ARGS
+      ${TribitsHelloWorld_COMMON_CONFIG_ARGS}
+      -DTribitsHelloWorld_ENABLE_TESTS=ON
+      -DTribitsHelloWorld_AUTOGENERATE_TEST_RESOURCE_FILE=ON
+      -DTribitsHelloWorld_CUDA_NUM_GPUS=2        # Will not get used!
+      -DTribitsHelloWorld_CUDA_SLOTS_PER_GPU=3   # Will not get used!
+      -DTPL_ENABLE_CUDA=ON
+      -DCTEST_RESOURCE_SPEC_FILE=other_ctest_resources.json
+      ${${PROJECT_NAME}_TRIBITS_DIR}/examples/TribitsHelloWorld
+    PASS_REGULAR_EXPRESSION_ALL
+      "NOTE: The test resource file CTEST_RESOURCE_SPEC_FILE='.*/TriBITS_TribitsHelloWorld_set_other_ctest_resources_file_autogen_on/other_ctest_resources[.]json' will not be auto-generated even through TribitsHelloWorld_AUTOGENERATE_TEST_RESOURCE_FILE=ON because its location does not match the default location '.*/TriBITS_TribitsHelloWorld_set_other_ctest_resources_file_autogen_on/ctest_resources[.]json'[.]  If you want to auto-generate this file, please clear CTEST_RESOURCE_SPEC_FILE and reconfigure or create that file on your own and clear TribitsHelloWorld_AUTOGENERATE_TEST_RESOURCE_FILE[.]"
+      "Configuring done"
+    ALWAYS_FAIL_ON_NONZERO_RETURN
+
+  TEST_2
```

Comment:
Shouldn't this have `ALWAYS_FAIL_ON_NONZERO_RETURN`?

Reply:
@KyleFromKitware,

No, the default pass/fail criteria is for a zero return code what is the case here.

The options the `ALWAYS_FAIL_ON_NONZERO_RETURN` and `ALWAYS_FAIL_ON_ZERO_RETURN` were added to the `TRIBITS_ADD_ADVANCED_TEST()` function to address the case where you want to determine pass/fail based on a regex **and** you want to ensure the return code is zero or non-zero.   See:

* https://tribits.org/doc/TribitsDevelopersGuide.html#test-case-pass-fail-tribits-add-advanced-test

Basic CMake/CTest does not support an option like.  You have to either check the STDOUT or the return code but you can't check both.  We have found that important defects slip though if you don't check both the STDOUT and the return code.   CTest should really add options like  `ALWAYS_FAIL_ON_NONZERO_RETURN` and `ALWAYS_FAIL_ON_ZERO_RETURN`  to be complete (or you force developers to write their own test runners to be safe).

Does that make sense?

Reply:
Yes, thanks for the explanation.